File: client/src/services/attendance.service.js
import api from "~/utils/axios";

export const checkInFaceApi = async ({ accountId, imageBlob, latitude, longitude }) => {
  try {
    const formData = new FormData();
    formData.append("accountId", accountId);
    formData.append("image", imageBlob);
    formData.append("latitude", latitude);
    formData.append("longitude", longitude);

    const res = await api.post("/attendance/check-in", formData, {
      headers: { "Content-Type": "multipart/form-data" },
    });

    return res.data; // dữ liệu attendance object hoặc lỗi server
  } catch (error) {
    if (error.response) {
      throw error.response.data; // lỗi từ server
    }
    throw error;
  }
};

export const checkOutFaceApi = async ({ accountId, imageBlob }) => {
  try {
    const formData = new FormData();
    formData.append("accountId", accountId);
    formData.append("image", imageBlob);

    const res = await api.post("/attendance/check-out", formData, {
      headers: { "Content-Type": "multipart/form-data" },
    });

    return res.data; // attendance object sau check-out thành công
  } catch (error) {
    if (error.response) {
      throw error.response.data;
    }
    throw error;
  }
};


export const getMyAttendancePageApi = async (payload) => {
  try {
    const res = await api.post("/attendance/get-my-attendance", payload);
    return res.data;
  } catch (error) {
    if (error.response) throw error.response.data;
    throw error;
  }
};

export const getAttendanceByIdApi = async (attendanceId) => {
  try {
    const res = await api.get(`/attendance/${attendanceId}`);
    return res.data;
  } catch (error) {
    if (error.response) throw error.response.data;
    throw error;
  }
};


export const getAttendanceTodayStatusApi = async (accountId) => {
  try {
    const res = await api.get("/attendance/today-status", {
      params: { accountId },
    });
    return res.data; // "NOT_CHECKED_IN", "CHECKED_IN", "CHECKED_OUT"
  } catch (error) {
    if (error.response) {
      throw error.response.data;
    }
    throw error;
  }
};

export const getMissingCheckOutApi = async (fromDate, toDate) => {
  try {
    const params = {};
    if (fromDate) params.fromDate = fromDate; // yyyy-MM-dd
    if (toDate) params.toDate = toDate;       // yyyy-MM-dd

    const res = await api.get("/attendance/missing-checkout", { params });
    return res.data; // mảng Attendance
  } catch (error) {
    if (error.response) throw error.response.data;
    throw error;
  }
};

// Nhân viên gửi note giải trình cho thiếu check-out
export const submitMissingCheckOutNoteApi = async (attendanceId, note) => {
  try {
    const params = new URLSearchParams();
    params.append("attendanceId", attendanceId);
    params.append("note", note);

    const res = await api.post("/attendance/submit-missing-checkout-note", params, {
      headers: { "Content-Type": "application/x-www-form-urlencoded" }
    });
    return res.data; // Attendance đã update note
  } catch (error) {
    if (error.response) throw error.response.data;
    throw error;
  }
};

// HR duyệt hoặc từ chối giải trình thiếu check-out
export const resolveMissingCheckOutApi = async (attendanceId, note, approved) => {
  try {
    const params = new URLSearchParams();
    params.append("attendanceId", attendanceId);
    params.append("note", note);
    params.append("approved", approved);

    const res = await api.post("/attendance/resolve-missing-checkout", params, {
      headers: { "Content-Type": "application/x-www-form-urlencoded" }
    });
    return res.data; // Attendance sau khi duyệt hoặc từ chối
  } catch (error) {
    if (error.response) throw error.response.data;
    throw error;
  }
};

File: mobile/lib/services/attendance_service.dart
import 'dart:io';
import 'package:dio/dio.dart';
import 'package:mobile/services/api_service.dart';
import 'package:mobile/models/attendance_model.dart';
import 'package:http_parser/http_parser.dart';
import 'package:intl/intl.dart';
class AttendanceService {
  // CHECK-IN: gửi đúng key 'image'
  static Future<Attendance> checkInFace({
    required int accountId,
    required File imageFile,
    required double latitude,
    required double longitude,
  }) async {
    final formData = FormData.fromMap({
      'accountId': accountId,
      'image': await MultipartFile.fromFile(
        imageFile.path,
        filename: 'live.jpg',
        contentType: MediaType('image', 'jpeg'),
      ),
      'latitude': latitude,
      'longitude': longitude,
    });

    try {
      final res = await ApiService.client.post(
        '/attendance/check-in',
        data: formData,
        options: Options(validateStatus: (c) => c != null && c < 500),
      );

      if (res.statusCode != 200 && res.statusCode != 201) {
        final msg = (res.data is Map ? res.data['message']?.toString() : null)
            ?? 'Check-in failed (${res.statusCode})';
        throw msg;
      }

      final data = res.data is Map && res.data['data'] != null ? res.data['data'] : res.data;
      return Attendance.fromJson(data);
    } on DioException catch (e) {
      final msg = e.response?.data is Map ? e.response?.data['message']?.toString() : e.message;
      throw msg ?? 'Network error';
    }
  }

  // CHECK-OUT: cũng gửi 'image' + lat/long
  static Future<Attendance> checkOutFace({
    required int accountId,
    required File imageFile,
    required double latitude,
    required double longitude,
  }) async {
    final formData = FormData.fromMap({
      'accountId': accountId,
      'image': await MultipartFile.fromFile(
        imageFile.path,
        filename: 'live.jpg',
        contentType: MediaType('image', 'jpeg'),
      ),
      'latitude': latitude,
      'longitude': longitude,
    });

    try {
      final res = await ApiService.client.post(
        '/attendance/check-out',
        data: formData,
        options: Options(validateStatus: (c) => c != null && c < 500),
      );

      if (res.statusCode != 200 && res.statusCode != 201) {
        final msg = (res.data is Map ? res.data['message']?.toString() : null)
            ?? 'Check-out failed (${res.statusCode})';
        throw msg;
      }

      final data = res.data is Map && res.data['data'] != null ? res.data['data'] : res.data;
      return Attendance.fromJson(data);
    } on DioException catch (e) {
      final msg = e.response?.data is Map ? e.response?.data['message']?.toString() : e.message;
      throw msg ?? 'Network error';
    }
  }

  // Lấy danh sách attendance cá nhân (phân trang)
  static Future<(List<Attendance>, int)> getMyAttendancePage({
    int page = 1,
    int pageSize = 10,
    String? statusFilter,
    String? searchTerm,
    String? sortBy = "desc",
    DateTime? fromDate,
    DateTime? toDate,
  }) async {
    final payload = {
      "pageNumber": page,
      "pageSize": pageSize,
      "statusFilter": statusFilter,
      "searchTerm": searchTerm,
      "sortBy": sortBy,
      "fromDate": fromDate != null ? DateFormat('yyyy-MM-dd').format(fromDate) : null,
      "toDate": toDate != null ? DateFormat('yyyy-MM-dd').format(toDate) : null,
    };

    final response = await ApiService.client.post(
      '/attendance/get-my-attendance',
      data: payload,
    );

    final data = response.data['data'];
    final list = data['attendances'] as List;
    final totalPage = data['totalPage'] as int;

    final result = list.map((json) => Attendance.fromJson(json)).toList();

    return (result, totalPage);
  }

  // Lấy chi tiết chấm công theo ID
  static Future<Attendance> getAttendanceById(int attendanceId) async {
    final response = await ApiService.client.get('/attendance/$attendanceId');
    return Attendance.fromJson(response.data);
  }

  // Lấy trạng thái chấm công trong ngày
  static Future<String> getTodayStatus(int accountId) async {
    final response = await ApiService.client.get(
      '/attendance/today-status',
      queryParameters: {'accountId': accountId},
    );
    return response.data; // Trả về: "NOT_CHECKED_IN", "CHECKED_IN", "CHECKED_OUT"
  }

  // Lấy danh sách thiếu check-out theo ngày
  static Future<List<Attendance>> getMissingCheckOut({String? fromDate, String? toDate}) async {
    final query = {
      if (fromDate != null) 'fromDate': fromDate,
      if (toDate != null) 'toDate': toDate,
    };

    final response = await ApiService.client.get(
      '/attendance/missing-checkout',
      queryParameters: query,
    );

    final List data = response.data;
    return data.map((json) => Attendance.fromJson(json)).toList();
  }

  // Gửi ghi chú giải trình thiếu check-out
  static Future<Attendance> submitMissingCheckOutNote({
    required int attendanceId,
    required String note,
  }) async {
    final form = FormData.fromMap({
      'attendanceId': attendanceId.toString(),
      'note': note,
    });

    final response = await ApiService.client.post(
      '/attendance/submit-missing-checkout-note',
      data: form,
      options: Options(contentType: Headers.formUrlEncodedContentType),
    );

    return Attendance.fromJson(response.data);
  }

  // HR duyệt hoặc từ chối giải trình thiếu check-out
  static Future<Attendance> resolveMissingCheckOut({
    required int attendanceId,
    required String note,
    required bool approved,
  }) async {
    final form = FormData.fromMap({
      'attendanceId': attendanceId.toString(),
      'note': note,
      'approved': approved.toString(),
    });

    final response = await ApiService.client.post(
      '/attendance/resolve-missing-checkout',
      data: form,
      options: Options(contentType: Headers.formUrlEncodedContentType),
    );

    return Attendance.fromJson(response.data);
  }
}
